server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -1347,10 +1365,81 @@ public Builder copy() {
         return new Builder(this);
     }
 
+    public record DataStreamIndices(
+        String namePrefix,
+        List<Index> indices,
+        boolean rolloverOnWrite,
+        @Nullable DataStreamAutoShardingEvent autoShardingEvent
+    ) {
+
+        public Index getWriteIndex() {
+            return indices.get(indices.size() - 1);
+        }
+
+        public String provideName(String dataStreamName, long generation, long epochMillis) {
+            return getDefaultIndexName(namePrefix, dataStreamName, generation, epochMillis);
+        }
+
+        public static Builder backingBuilder(List<Index> indices) {
+            return new Builder(BACKING_INDEX_PREFIX, indices);
+        }
+
+        public static Builder failureBuilder(List<Index> indices) {
+            return new Builder(FAILURE_STORE_PREFIX, indices);
+        }
+
+        public Builder copy() {
+            return new Builder(this);
+        }
+
+        public static class Builder {
+            private String namePrefix;
+            private List<Index> indices;
+            private boolean rolloverOnWrite = false;
+            @Nullable
+            private DataStreamAutoShardingEvent autoShardingEvent = null;
+
+            public Builder(String namePrefix, List<Index> indices) {
+                this.namePrefix = namePrefix;
+                this.indices = indices;
+            }
+
+            public Builder(DataStreamIndices dataStreamIndices) {
+                this.namePrefix = dataStreamIndices.namePrefix;
+                this.indices = dataStreamIndices.indices;
+                this.rolloverOnWrite = dataStreamIndices.rolloverOnWrite;
+                this.autoShardingEvent = dataStreamIndices.autoShardingEvent;
+            }
+
+            public Builder setNamePrefix(String namePrefix) {
+                this.namePrefix = namePrefix;
+                return this;
+            }
```

Comment:
Does it make sense to make this modifiable outside of the first time making the builder?

Reply:
Didn't necessarily see an immediate reason to forbid that, but I'm also fine with removing the setter. Will do

Reply:
Okay nvm, there are actually use cases where we copy a data stream and then change its name, for example [here](https://github.com/nielsbauman/elasticsearch/blob/data-stream-store-extraction/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L708-L708).

Reply:
That sets the data stream name, but that shouldn't have any effect on the prefix that we use for internal indices right?